### ISSUE TITLE
Question: Pipeline Queue with full serial processing

### ISSUE THREAD

**Issue by StijnBt** [HodorNV/ALOps#364]
Hello,

I don't think it's more a general question and not so much related to ALOps per se.
But I have a pipeline that contains 4 stages: build, release test, release accp & release prod.
The 4 stages are handled by 2 templates (build & release).
The build templates holds 4 jobs with each their specific tasks/steps.
So far so good.

Now becomes the tricky part. Let's assume I have 20 repos that I want to build to NextMinor & NextMajor.
Since I use generic templates & pipelines all of them have a cron job that starts at midnight 2 times a month.
Thats 40 builds launched... and I assumed nicely queued. The truth however is that some stages from one pipeline are launched and all others from that same pipeline are put in hold and another new pipeline is launched and again and again and ...
So instead of just handling one pipeline completely and then switching over to the next one, I end up with a lot of them running in some parallel mode (lots of dockers running!) and things start failing all over the place.

Is there a way to just force a build agent to queue up jobs and process them in a pure serial mode.
I've been googeling quite a while and found no solution so far, I found an item that suggests it isn't possible... but that's just too hard to believe.

**Comment by waldo1001** [comment]
So you have multiple agents on one machine?

**Comment by StijnBt** [comment]
No just the "one"

![image](https://user-images.githubusercontent.com/37275572/116580467-b02acd80-a913-11eb-8def-94c2e5e4ff78.png)

If I have this pipeline, and gets started for 3 repos at the same time.
The it will execute as follows:
Repo1-Build & compile app-Pre actions
Repo2-Build & compile app-Pre actions
Repo3-Build & compile app-Pre actions
Repo1-Build & compile app-Create Docker
Repo2-Build & compile app-Create Docker
Repo3-Build & compile app-Create Docker
Repo1-Build & compile app-Build &compile
Repo2-Build & compile app-Build &compile
Repo3-Build & compile app-Build &compile
.....

Where I was hoping for:
Repo1-Build & compile app-Pre actions
Repo1-Build & compile app-Create Docker
Repo1-Build & compile app-Build &compile
Repo1-Build & compile app-Post actions
Repo1-Build & compile dependant-Pre actions
Repo1-Build & compile dependant-Create Docker
Repo1-Build & compile dependant-Build &compile
Repo1-Build & compile dependant-Post actions
Repo1-Build & compile dependant-Deploy to test
Repo2-Build & compile app-Pre actions
Repo2-Build & compile app-Create Docker
Repo2-Build & compile app-Build &compile
Repo2-Build & compile app-Post actions
Repo2-Build & compile dependant-Pre actions
Repo2-Build & compile dependant-Create Docker
Repo2-Build & compile dependant-Build &compile
Repo2-Build & compile dependant-Post actions
Repo2-Build & compile dependant-Deploy to test
.....

I think it's this, not sure: 
https://developercommunity.visualstudio.com/t/prevent-parallel-execution-of-the-same-build-defin/365730?from=email&viewtype=all#T-ND1410901

**Comment by waldo1001** [comment]
Just a guess - but did you use the "dependsOn" in your pipeline for the next stage to wait for the previous stage?
![image](https://user-images.githubusercontent.com/12088142/117304119-aeb25580-ae7d-11eb-968f-7c6389663618.png)

**Comment by waldo1001** [comment]
And?

**Comment by StijnBt** [comment]
Well... I guess it's possible according to [this](https://stackoverflow.com/questions/65401974/devops-pipeline-use-dynamic-parameters-for-job-name) post on stackoverflow.

But passing a parameter to the template is what I want to avoid, because then I'll need to modify all pipelines yamls calling this template. And by now that are a lot of them.

So I tried to achieve the same thing with some of the [predefined variables](https://docs.microsoft.com/en-us/azure/devops/pipelines/build/variables?view=azure-devops&tabs=yaml) directly in my template, yet with no luck.
So I decided to restructure my template and combine the "CreateDocker" & "Build & compile" job in a single job, since it were those who were causing the main issue (lots of running containers on build server).

So my problem is solved, but not exactly "how" I had in mind.
